paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CommitterOperator.java

```diff
@@ -188,28 +189,37 @@ public void endInput() throws Exception {
         }
 
         pollInputs();
-        commitUpToCheckpoint(Long.MAX_VALUE);
+        commitUpToCheckpoint(END_INPUT_CHECKPOINT_ID);
     }
 
     @Override
     public void notifyCheckpointComplete(long checkpointId) throws Exception {
         super.notifyCheckpointComplete(checkpointId);
-        commitUpToCheckpoint(endInput ? Long.MAX_VALUE : checkpointId);
+        commitUpToCheckpoint(endInput ? END_INPUT_CHECKPOINT_ID : checkpointId);
     }
 
     private void commitUpToCheckpoint(long checkpointId) throws Exception {
         NavigableMap<Long, GlobalCommitT> headMap =
                 committablesPerCheckpoint.headMap(checkpointId, true);
         List<GlobalCommitT> committables = committables(headMap);
-        committer.commit(committables);
-        headMap.clear();
+        if (committables.isEmpty() && committer.forceCreatingSnapshot()) {
+            committables =
+                    Collections.singletonList(
+                            toCommittables(checkpointId, Collections.emptyList()));
+        }
 
-        if (committables.isEmpty()) {
-            if (committer.forceCreatingSnapshot()) {
-                GlobalCommitT commit = toCommittables(checkpointId, Collections.emptyList());
-                committer.commit(Collections.singletonList(commit));
-            }
+        if (checkpointId == END_INPUT_CHECKPOINT_ID) {
+            // In new versions of Flink, if a batch job fails, it might restart from some operator
+            // in the middle.
+            // If the job is restarted from the commit operator, endInput will be called again, and
+            // the same commit messages will be committed again.
+            // So when `endInput` is called, we must check if these commit messages are already
+            // committed.
+            committer.filterAndCommit(committables);
```

Comment:
Here should filter commit, but should not check add files. The latter has a relatively high cost.